### PR TITLE
Spec 2: Election UI (initial draft based on google doc)

### PR DESCRIPTION
**Pull request by starpause**

### CODE REVIEW

File: specs/spec-2-election-ui.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+The following is to provide context for the electing the NDC Congressional Bodies, which are the House of Merit, Transparency Commission, and Council of Advisors. Voting is conducted by verified humans in the NDC (denoted as members of the Voting Body).
+
+The overall governance flow:
+
+1. **Nomination process**, which involves a self vote (based off of kudos) and accompanying attestation and a kudo + attestation by another voting body members
+2. **Election process** (the scope of this RFP), which involves voting body members having x amount of credits per congressional body and congressional 
+3. **Congress in session**: when the elected congressional members make decisions during their bi-annual tenure in their body including passing setup packages, approving budget of constellations etc. 
+
+We are assuming Nomination->Election ->Congressional Onboarding are done manually, (ex; by the governance working group) as opposed to automatically updated by cross contract calls. Each part of the process which is not manually handled will be on-chain. Automation will be worth building once we've proven the system works via manual effort.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:33]:
```suggestion
We are assuming stage changes: Nomination -> Election -> Congressional Onboarding are done manually, (ex; by the governance working group) as opposed to automatically updated by cross contract calls. Each part of the process, which is not manually handled, will be on-chain. Automation will be worth building once we've proven the system works via manual effort. 
```

File: specs/spec-2-election-ui.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+Canidates are included in the Election UI after they've been nominated via [Kudos](https://github.com/near-ndc/gwg/blob/main/specs/spec-1-kudos.md) (or a Nominate specific fork of Kudos). After an election the winning candidates are assigned seats in NDC congress.
+
+## Motivation
+
+<!-- Context one needs to know to understand this doc: motivating examples, previous versions and problems, links to related projects/design docs, etc. You should mention related work outside of NEAR if applicable. Note: this is background; do not write about your design or ideas to solve problems here. What's the "historical" context? Why do we need a new project? Identify the problems and the needs. -->
+
+The following is to provide context for the electing the NDC Congressional Bodies, which are the House of Merit, Transparency Commission, and Council of Advisors. Voting is conducted by verified humans in the NDC (denoted as members of the Voting Body).
+
+The overall governance flow:
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:27]: should we include in the process a step of candidate presentation (thesis, profile etc...)?

File: specs/spec-2-election-ui.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+- Watchdog (no IAH credentials)
+
+These user stories illustrate the needs and desires of the actors rather than perscribing how the actors will achieve their goals.
+
+### Voter
+
+#### Outside of Election UI but still in BOS
+
+- I want to be notified when the election opens and notified when the election is about to close directly on BOS [notification widget].
+- I want to know about upcoming elections and their dates.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:55]: Let's maybe specify that we are expecting a kind of a dashboard with:

* section about upcoming elections (should be broken down by a house? or all combined?)
* dashboard with latest news.

File: specs/spec-2-election-ui.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+- I want to see my past votes casted for each congressional body, so I know who I voted for.
+- I want to understand the time frame I'm looking at defined both in terms of Congressional # and year.
+- I want to see the results of an election after it was over with transaction history, so there is accountability after the election.
+
+#### Voting in a Live Election
+
+- I want a cart of candidates, so I can organize who I am voting for before I cast a vote on chain.
+- I want to be able to vote on each Congressional body one at a time so I don’t have to make the entire decision at once.
+- I want to be able to toggle between Congressional bodies so I can more easily choose candidates and my vote status.
+- I want to see how many votes I have cast and have left for each congressional body, to better allocate my votes and understand the quadratic voting process.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:71]:
```suggestion
- I want to see how many _votes_ I have cast and how many _voting credits_ I have left for each congressional body, to better allocate my votes and understand the quadratic voting process.
```

File: specs/spec-2-election-ui.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+- I want to learn about quadratic voting because it's new to me.
+- I want to easily increment and decrement the votes I've assigned to each canidate.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:73]:
```suggestion
- I want to easily increment and decrement the votes I've assigned to each canidate, and see the global voting credit change (based on quadratic voting).
```

Comment (robert-zaremba) [reply specs/spec-2-election-ui.md:73]: let's keep QV for the moment, and we will update it later once we have a decision.

File: specs/spec-2-election-ui.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+#### Voting in a Live Election
+
+- I want a cart of candidates, so I can organize who I am voting for before I cast a vote on chain.
+- I want to be able to vote on each Congressional body one at a time so I don’t have to make the entire decision at once.
+- I want to be able to toggle between Congressional bodies so I can more easily choose candidates and my vote status.
+- I want to see how many votes I have cast and have left for each congressional body, to better allocate my votes and understand the quadratic voting process.
+- I want to learn about quadratic voting because it's new to me.
+- I want to easily increment and decrement the votes I've assigned to each canidate. 
+- I want to know when I've used up all of my votes so I don't case more votes than I have (Disabled UI, toast notification)
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:74]:
```suggestion
- I want to know when I've used up all of my vote credits so I don't case more votes than I have (Disabled UI, toast notification)
```

File: specs/spec-2-election-ui.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+- I want to save my progress for voting/cart, so I can resume + calculate quadratic voting strategy at a later vote without having to submit the whole decision.
+- I want to see time until the next Congressional election after a vote is over, so I can plan to run for upcoming congressional elections.
+- I want to see past election results (even if its first vote - show no past election), so I can gauge the electorate’s sentiment.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:77]: Maybe we can move it to v2?

File: specs/spec-2-election-ui.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+#### Viewing Election Candidates
+
+- I want to see a candidate profile (Near Social) and the platform they used to nominate themselves (nomination mechism).
+- I want to see a candidate's past congressional history on profile (past roles, proposals, and complaints), so I can audit their past performance.
+- I want to see the profile + attestation (queried from the nomination process) for a candidate so I can see social proof for why someone was elected.
+- I want to see what DAOs a candidate is in even if they are not part of NDC, in order to understand what communities they are a part of.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:84]: We will need to specify it. How we will keep track of this? Are we only going to look at Astro DAO (only contract deployed by the factory), or other contracts as well?

File: specs/spec-2-election-ui.md [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+#### Runoffs
+
+- I want to know whether the election resulted in any tie, and what are the clear next steps to finalizing runoff.
+
+### Watchdog
+
+- I want to see details about the ongoing election to judge if it's worth my time and effort to join the NDC and be heard.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:92]: We also need to better specify this.
* how to enter the watchdog dashboard?
* who can do it? if it's going to be fully on chain, then everyone will be able to see it.
* who will assign watchdog and if we want to implement the assignment process?

File: specs/spec-2-election-ui.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+### Candidate
+
+- I want elections to be relatively short so that the time I spend campaigning can be focused. 
+- I want to see how much of the voting body has voted in total so I can plan campaign efforts.
+
+### Additional Features & Requirements
+
+<!-- Optionally, describe: Expected functionality, Security issues/requirements, Assumptions/dependencies (internal and external), Potential problems/roadblocks, Budget concerns. -->
+
+- Edit profile in the scope of near social profile
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:103]: What kind of profile? Is the existing profile widget good enough or you think we should have a new widget? Let's add more details.

File: specs/spec-2-election-ui.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+- Results -> green check mark
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:104]: where should the results be displayed? New page? Same page as elections (voting)?

File: specs/spec-2-election-ui.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+- Votes can not be changed once cast, but can be changed before they are.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:105]: let's clarify: we need a "submit" button. Users can allocate votes ....

File: specs/spec-2-election-ui.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+<!-- Does the widget interact with other widgets? -->
+
+- Kudos Widget
+- Near Social Profile
+
+## Open Questions and Comments
+
+<!-- Topics not covered in this document that need to be addressed in a meeting or async. Potential future improvements, Summary of open discussion. -->
+### Open
+
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:152]: let's add a note that we will also need an indexer. We should create a separate RFP (based on this spec) to provide indexer features:
* query candidates
* query votes
* query ongoing results.

Suggestion: let's add it in the section above under: Tech Spec / Dependent Backend

File: specs/spec-2-election-ui.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+  - **A:** Election timing will be retrieved from the smart contract. It will be set on the smart contract by NDC administrators.  
+- **Q:**Considering votes are on-chain and don’t lack privacy, should we be able to see votes? In the spec above we are assuming yes. Going off of this, we are unclear on whether a voting body member can override their vote for a congressional body before the deadline as it would make the vote even more susceptible to swindling.
+  - **A:** Yes, we will be displaying votes
+- **Q:** One of the biggest challenges for NDC Voting body members is to notify them of a vote. THis brings an overall question of outside of BOS Push notification, and what point of the I-AM-HUMAN onboarding flow will users optin and enable Web2 notifications (in a GDPR compliant fashion) so we can remind them of voting (and other NDC critical updates) outside of BOS. This may result in a separate spec for overall NDC GDPR Compliant Notifications
+  - **A:** We will use standard BOS notifications.
+- **Q:** What is our runoff policy? 
+  - **A:** An NDC administrator will begin a new runoff election in the case of a tie.
```
Comment (robert-zaremba) [inline specs/spec-2-election-ui.md:162]: We don't have NDC administrator. How need to define the role, how about Voting Comission, Voting Comissioner, or GWG (need to check in GWG about more opinions)

**Review by robert-zaremba — COMMENTED** [review]
No blocking comment, but let's add more clarifications before approving.

File: specs/spec-2-election-ui.md (hunk 2 above)
Comment (starpause) [reply specs/spec-2-election-ui.md:27]: MVP won't have user profile pages. Candidate names link to their NS profile, candidate "Nomination" links to Kudos/Nominate.

File: specs/spec-2-election-ui.md (hunk 3 above)
Comment (starpause) [reply specs/spec-2-election-ui.md:55]: MVP won't have a separate dashboard. Upcoming elections can be discovered two ways in the current interface: (1) Congress/Time selector (2) List of upcoming elections under the current election status graphic.

File: specs/spec-2-election-ui.md (hunk 11 above)
Comment (starpause) [reply specs/spec-2-election-ui.md:104]: Results are displayed on the same page as elections/voting.

File: specs/spec-2-election-ui.md (hunk 12 above)
Comment (starpause) [reply specs/spec-2-election-ui.md:105]: The wireframes have a "submit votes" button.

File: specs/spec-2-election-ui.md (hunk 13 above)
Comment (starpause) [reply specs/spec-2-election-ui.md:158]: These have been captured under ## Tech Spec ### Smart Contract Functions. Can we capture enough detail here that the developers working on this spec are responsible for figuring out how to get data from the smart contract? How do most BOS widgets get their information today? Can that same technique be used for Election UI?

File: specs/spec-2-election-ui.md (hunk 14 above)
Comment (starpause) [reply specs/spec-2-election-ui.md]: Starting elections is handled entirely on the contract side via CLI. The voting administrators will be a list of addresses on the voting contract who have permissions to call these functions.

File: specs/spec-2-election-ui.md (hunk 2 above)
Comment (robert-zaremba) [reply specs/spec-2-election-ui.md:27]: right, I was thinking to add a bit more details to the flow. Or ideally a link to external document where this is discussed.

File: specs/spec-2-election-ui.md (hunk 3 above)
Comment (robert-zaremba) [reply specs/spec-2-election-ui.md:55]: ok, then let's describe it in the spec.

File: specs/spec-2-election-ui.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+---
+
+# Spec-2: Election UI
+
+## Summary
+
+<!-- One page high-level overview; put details in the specification section and background in the previous section. Should be understandable by a new engineer or reader who isn't already working on the project. Short summary in layman terms. -->
+<!-- List the challenge(s) being solved by this widget. What is the problem? -->
+
+The Election UI provides an interface for people to see the status of current + past NDC Congressional Body elections. This includes the candidates and votes that have been cast.
```
[the remark this answers is not in the document]
Comment (starpause) [reply specs/spec-2-election-ui.md:17]: Addressed in ## Work Plan ### Sprints

**Review by robert-zaremba — APPROVED** [review]
pre-approving. Let's separate what's v1 and what's for v2 (or later).

**Review by starpause — COMMENTED** [review]
LGTM

**Review by starpause — COMMENTED** [review]
Resolved conversations.

**Review by starpause — COMMENTED** [review]
More resolved conversations.

**Review by robert-zaremba — APPROVED** [review]
